xarray/core/variable.py

```diff
@@ -940,7 +940,11 @@ def _shift_one_dim(self, dim, count):
             keep = slice(None)
 
         trimmed_data = self[(slice(None),) * axis + (keep,)].data
-        dtype, fill_value = dtypes.maybe_promote(self.dtype)
+
+        if fill_value is None or fill_value is np.nan:
```

Comment:
Use `dtypes.NA` as the default value for `fill_value`, and then copy these lines from `pad` to ensure that this works for arbitrary fill_values such as `None`:
https://github.com/pydata/xarray/blob/master/xarray/core/variable.py#L1012-L1015

Reply:
I saw that, but it means that if someone passes `np.nan` (the value, not the `dtypes.NA` default) to a int dtype, I get `-9223372036854775808`. Should we raise in that case? Or I'll see what I can do to coerce.

Reply:
> I'm mostly concerned with duplicate logic that works slightly differently
💯 

Great, I'll have a look and report back